src/Microsoft.ML.Vision/ImageClassificationTrainer.cs

```diff
@@ -752,15 +752,15 @@ private void CheckTrainingParameters(Options options)
             var decodedImage4d = tf.expand_dims(decodedImageAsFloat, 0);
             var resizeShape = tf.stack(new int[] { inputDim.Item1, inputDim.Item2 });
             var resizeShapeAsInt = tf.cast(resizeShape, dtype: tf.int32);
-            var resizedImage = tf.image.resize_bilinear(decodedImage4d, resizeShapeAsInt, false, "ResizeTensor");
+            var resizedImage = tf.image.resize_bilinear(decodedImage4d, resizeShapeAsInt, false, name: "ResizeTensor");
             return (jpegData, resizedImage);
         }
 
         private static Tensor EncodeByteAsString(VBuffer<byte> buffer)
         {
             int length = buffer.Length;
-            var size = c_api.TF_StringEncodedSize((UIntPtr)length);
-            var handle = c_api.TF_AllocateTensor(TF_DataType.TF_STRING, IntPtr.Zero, 0, (UIntPtr)((ulong)size + 8));
+            var size = c_api.TF_StringEncodedSize((ulong)length);
+            var handle = c_api.TF_AllocateTensor(TF_DataType.TF_STRING, new long[0], 0, ((ulong)size + 8));
```

Comment:
Do you have to create a new, empty array here every time? Instead, would it be possible to use `Array.Empty<long>()`, so it doesn't allocate and throw away such a small object?

Reply:
Good point

Reply:
What's the difference between `Array.Empty<long>()`, `new long[0]` and `(long*)(IntPtr.Zero)`, especially why using `Array.Empty<long>()` can avoid extra mem allocate.

Reply:
`new long[0]` will always create a new managed array object. So new memory (to store the object) will always be allocated every time this method is called. And when the code is done with each object, they will be garbage collected.

`Array.Empty<long>()` on the other hand will allocate a single empty array for the entire process, and return that each time. See https://github.com/dotnet/runtime/blob/9c472545df220d7ae26329e1649f880c4ee127e8/src/libraries/System.Private.CoreLib/src/System/Array.cs#L702-L712.

`(long*)(IntPtr.Zero)` is completely different in that it isn't a managed array at all - it is a `nullptr` to a `long`.